### PR TITLE
Build Tool Search and Acquire Experience

### PR DESCRIPTION
**Pull request by ravimeda**
Currently, if a prerequisite for a build is missing, then the build fails. This experience can be improved if the missing prerequisite is acquired, and made available for the build to consume.

Summary of changes:
1. Metadata of each required build tool will be in `.toolversions` file located in repository root
2. Native build invokes a probing mechanism to get a required tool. Build should specify the name of the required tool. Optionally build can also specify if the declared version, which the version that has been verified to work correctly with the product, of the tool is required. Probing mechanism invokes search and acquire scripts to get the tool on behalf of the build, and returns the tool path if successful or an error message. 
3. Scripts that search and acquire a tool will be in `tools-local` folder in repository root. 
Short description of the design is provided below. Implementation of this design for probing CMake is ready.

The design of these scripts is like having ISearchTool and IAcquireTool interfaces. ISearchTool provides a base implementation of searching for the specified tool in environment path, and the search paths specified in the tools metadata file. Base implementation is in `tools-local/unix/search-tool.sh`, and a corresponding PowerShell script for Windows. A tool can override this implementation if it wants to follow a customized search process. This override script will go into the tool specific folder. For example, if CMake overrides the base implementation then, the override script should be in `tools-local/unix/cmake/search-tool.sh`. 

ISearchTool has abstract methods that each tool must provide an implementation. For example, `get-version.sh` gets the version of CMake at the specified path. This script will be in tool specific folder at `tools-local/unix/cmake/get-version.sh`.

IAcquireTool provides similar virtual and abstract functions for acquiring a tool.
Entry point to these scripts is `probe-tool.sh` that gets invoked by `src/Native/build-native.sh`.

In the current implementation, a tool is downloaded directly from the internet. If a developer would like to acquire the tool from a different location, then the developer can either modify `tools-local/unix/acquire-tool.sh` within the local repository or have the customized script in a folder rooted outside the repository.  Path to the folder containing such customized scripts can be specified using `OverrideScriptsFolderPath` argument. A future addition would be to provide an option to override the default `.toolversions` file.

@weshaggard @ellismg PTAL.

### CODE REVIEW

**Comment by weshaggard** [comment]
> using OverrideScriptsFolderPath argument. A future addition would be to provide an option to override the default .toolversions file.

IIRC after our last review we decided to not use OverrideScritpsFolderPath and instead override .toolsversions? What is our plan for official builds with the current approach?

File: tools-local/windows/tool-helper.ps1 [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    [CmdletBinding()]
+    param(
+        [parameter(Mandatory=$true, Position=0)]
+        [string]$ToolName,
+        [ValidateNotNullOrEmpty()]
+        [parameter(Mandatory=$true, Position=1)]
+        [string]$RepositoryRoot
+    )
+
+    $toolVersionsFilePath = Join-Path "$RepositoryRoot" ".toolversions"
```
Comment (weshaggard) [inline tools-local/windows/tool-helper.ps1]: I hope you are planning to cache this in some way. Currently everytime Get-ToolVersionsConfig or transitiviely Get-ToolConfigValue is called you re-read the file from disk to parse the value. We should read the tools version file once.

Comment (ravimeda) [inline tools-local/windows/tool-helper.ps1]: > What is our plan for official builds with the current approach?

Yes, the plan for official builds is still to use a copy of `.toolversions` that overrides the default one. The path to this file will be passed as a command line argument. I'm thinking we can place that file (not all scripts, just the file only) in a folder whose path can be passed though `OverrideScritpsFolderPath` , and modify `tool-helper` to consume the config values listed in that override file.

**Comment by ravimeda** [comment]
Please ignore recent commits `ae20cd7'.  I will let you know when PR is ready for review.

**Comment by karelz** [comment]
Closing the PR. @ravimeda please reopen it once it is ready. Thanks!

**Pull request closed without merging**